hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java

```diff
@@ -8905,69 +8683,44 @@ public RetryCacheCleaner() {
           DFS_NAMENODE_RETRY_CACHE_EXPIRYTIME_MILLIS_DEFAULT);
     }
 
+    private int deleteAllForEpoch(final long epoch) throws IOException {
+      return (int)(new LightWeightRequestHandler(
+              HDFSOperationType.CLEAN_RETRY_CACHE) {
+        @Override
+        public Object performTask() throws IOException {
+
+          RetryCacheEntryDataAccess da = (RetryCacheEntryDataAccess) HdfsStorageFactory
+                  .getDataAccess(RetryCacheEntryDataAccess.class);
+          return da.removeOlds(epoch);
+        }
+      }.handle());
+    }
+
     @Override
     public void run() {
-      int numRun = 0;
-      while (fsRunning && shouldCacheCleanerRun) {
-        try {
-          long start = monotonicNow();
-          final List<CacheEntry> toRemove = new ArrayList<>();
-          int num = retryCache.getToRemove().drainTo(toRemove);
-          if (num > 0) {
-            HopsTransactionalRequestHandler rh
-                = new HopsTransactionalRequestHandler(HDFSOperationType.CLEAN_RETRY_CACHE) {
-              @Override
-              public void acquireLock(TransactionLocks locks) throws IOException {
-                LockFactory lf = getInstance();
-                locks.add(lf.getRetryCacheEntryLock(toRemove));
-              }
-
-              @Override
-              public Object performTask() throws IOException {
-                int i = 0;
-                Iterator<CacheEntry> it = toRemove.iterator();
-                while (it.hasNext() && i<100) {
-                  CacheEntry entry = it.next();
-                  if (EntityManager.find(RetryCacheEntry.Finder.ByClientIdAndCallId, entry.getClientId(), entry.
-                      getCallId()) != null) {
-                    EntityManager.remove(new RetryCacheEntry(entry.getClientId(), entry.getCallId()));
-                  }
-                  i++;
-                  it.remove();
-                }
-                return null;
+      try{
+          while (fsRunning && shouldCacheCleanerRun ) {
```

Comment:
I think the try catch should be in the while, otherwise, some transient error in the DB may result in the cleaner thread terminating.